smarts/core/agent_manager.py

```diff
@@ -348,15 +351,46 @@ def _filter_social_agent_actions_for_controlled_vehicles(
 
         return social_agent_actions
 
-    def send_observations_to_social_agents(self, observations):
-        """Forwards observations to managed social agents."""
-        # TODO: Don't send observations (or receive actions) from agents that have done
-        #       vehicles.
+    def add_social_observation_callback(self, callback, id):
+        """Suscribe to observe social agents."""
+        self._social_agent_observation_callbacks[id] = callback
+
+    def remove_social_observation_callback(self, id):
+        """Remove a subscription to social agents."""
+        del self._social_agent_observation_callbacks[id]
+
+    def reserve_social_agent_action(self, agent_id, action):
+        """Override a current social agent action."""
+        self._reserved_social_agent_actions[agent_id] = action
+
+    def _send_observations_to_social_agents(self, observations: Dict[str, Observation]):
         self._remote_social_agents_action = {}
+        for agent_id, action in self._reserved_social_agent_actions.items():
+            future_action = futures.Future()
+            future_action.set_result(action)
+            self._remote_social_agents_action[agent_id] = future_action
+        self._reserved_social_agent_actions.clear()
+        for callback in self._social_agent_observation_callbacks.values():
+            callback(
+                dict(
+                    filter(
+                        lambda k: k[0] in self._remote_social_agents,
+                        observations.items(),
```

Comment:
```suggestion
                        lambda k: k in self._remote_social_agents,
                        observations.keys(),
```

Reply:
Sorry, no the original was correct because it was to do a filter on the the items and return them as pairs for the construction of the dictionary. Using `keys()` means there are only keys for the reconstruction.